File: src/main/java/com/demonwav/mcdev/platform/forge/inspections/sideonly/LocalVariableDeclarationSideOnlyInspection.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 *
 * MIT License
 */
package com.demonwav.mcdev.platform.forge.inspections.sideonly;

import com.demonwav.mcdev.util.McPsiUtil;

import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiClassType;
import com.intellij.psi.PsiLocalVariable;
import com.intellij.psi.PsiModifierListOwner;
import com.intellij.psi.PsiType;
import com.siyeh.ig.BaseInspection;
import com.siyeh.ig.BaseInspectionVisitor;
import com.siyeh.ig.InspectionGadgetsFix;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class LocalVariableDeclarationSideOnlyInspection extends BaseInspection {

    @Nls
    @NotNull
    @Override
    public String getDisplayName() {
        return "Invalid usage of local variable declaration annotated with @SideOnly";
    }

    @NotNull
    @Override
    protected String buildErrorString(Object... infos) {
        final Error error = (Error) infos[0];
        return error.getErrorString(SideOnlyUtil.getSubArray(infos));
    }

    @Nullable
    @Override
    public String getStaticDescription() {
        return "A variable whose class declaration is annotated with @SideOnly for one side cannot be declared in a class" +
            " or method that does not match the same side.";
    }

    @Nullable
    @Override
    protected InspectionGadgetsFix buildFix(Object... infos) {
        final PsiClass variableClass = (PsiClass) infos[3];

        if (variableClass.isWritable()) {
            return new RemoveAnnotationInspectionGadgetsFix() {
                @Nullable
                @Override
                public PsiModifierListOwner getListOwner() {
                    return variableClass;
                }

                @Nls
                @NotNull
                @Override
                public String getName() {
                    return "Remove @SideOnly annotation from variable class declaration";
                }
            };
        } else {
            return null;
        }
    }

    @Override
    public BaseInspectionVisitor buildVisitor() {
        return new BaseInspectionVisitor() {
            @Override
            public void visitLocalVariable(PsiLocalVariable variable) {
                final PsiClass psiClass = McPsiUtil.getClassOfElement(variable);
                if (psiClass == null) {
                    return;
                }

                if (!SideOnlyUtil.beginningCheck(variable)) {
                    return;
                }

                final PsiType type = variable.getType();

                if (!(type instanceof PsiClassType)) {
                    return;
                }

                final PsiClassType classType = (PsiClassType) type;

                final PsiClass variableClass = classType.resolve();
                if (variableClass == null) {
                    return;
                }

                final Side variableSide = SideOnlyUtil.getSideForClass(variableClass);
                if (variableSide == Side.NONE || variableSide == Side.INVALID) {
                    return;
                }

                final Side containingClassSide = SideOnlyUtil.getSideForClass(psiClass);
                final Side methodSide = SideOnlyUtil.checkElementInMethod(variable);

                boolean classAnnotated = false;

                if (containingClassSide != Side.NONE && containingClassSide != Side.INVALID) {
                    if (variableSide != containingClassSide) {
                        registerVariableError(
                            variable,
                            Error.VAR_CROSS_ANNOTATED_CLASS,
                            variableSide.getName(),
                            containingClassSide.getName(),
                            variableClass
                        );
                    }
                    classAnnotated = true;
                }

                if (methodSide == Side.INVALID) {
                    return;
                }

                if (variableSide != methodSide) {
                    if (methodSide == Side.NONE) {
                        if (!classAnnotated) {
                            registerVariableError(
                                variable,
                                Error.VAR_UNANNOTATED_METHOD,
                                variableSide.getName(),
                                methodSide.getName(),
                                variableClass
                            );
                        }
                    } else {
                        registerVariableError(
                            variable,
                            Error.VAR_CROSS_ANNOTATED_METHOD,
                            variableSide.getName(),
                            methodSide.getName(),
                            variableClass
                        );
                    }
                }
            }
        };
    }

    enum Error {
        VAR_CROSS_ANNOTATED_CLASS {
            @Override
            String getErrorString(Object... infos) {
                return "A local variable whose class is annotated with " + infos[0] + " cannot be used in a class annotated with " + infos[1];
            }
        },
        VAR_CROSS_ANNOTATED_METHOD {
            @Override
            String getErrorString(Object... infos) {
                return "A local variable whose class is annotated with " + infos[0] + " cannot be used in a method annotated with " + infos[1];
            }
        },
        VAR_UNANNOTATED_METHOD {
            @Override
            String getErrorString(Object... infos) {
                return "A local variable whose class is annotated with " + infos[0] + " cannot be used in an un-annotated method.";
            }
        };

        abstract String getErrorString(Object... infos);
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/forge/inspections/sideonly/LocalVariableDeclarationSideOnlyInspection.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.forge.inspections.sideonly

import com.demonwav.mcdev.util.findContainingClass
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiClassType
import com.intellij.psi.PsiLocalVariable
import com.siyeh.ig.BaseInspection
import com.siyeh.ig.BaseInspectionVisitor
import com.siyeh.ig.InspectionGadgetsFix
import org.jetbrains.annotations.Nls

class LocalVariableDeclarationSideOnlyInspection : BaseInspection() {

    @Nls
    override fun getDisplayName() = "Invalid usage of local variable declaration annotated with @SideOnly"

    override fun buildErrorString(vararg infos: Any): String {
        val error = infos[0] as Error
        return error.getErrorString(*SideOnlyUtil.getSubArray(infos))
    }

    override fun getStaticDescription() =
        "A variable whose class declaration is annotated with @SideOnly for one side cannot be declared in a class" +
            " or method that does not match the same side."

    override fun buildFix(vararg infos: Any): InspectionGadgetsFix? {
        val variableClass = infos[3] as PsiClass

        return if (variableClass.isWritable) {
            RemoveAnnotationInspectionGadgetsFix(
                variableClass,
                "Remove @SideOnly annotation from variable class declaration"
            )
        } else {
            null
        }
    }

    override fun buildVisitor(): BaseInspectionVisitor {
        return object : BaseInspectionVisitor() {
            override fun visitLocalVariable(variable: PsiLocalVariable) {
                val psiClass = variable.findContainingClass() ?: return

                if (!SideOnlyUtil.beginningCheck(variable)) {
                    return
                }

                val type = variable.type as? PsiClassType ?: return

                val variableClass = type.resolve() ?: return

                val variableSide = SideOnlyUtil.getSideForClass(variableClass)
                if (variableSide === Side.NONE || variableSide === Side.INVALID) {
                    return
                }

                val containingClassSide = SideOnlyUtil.getSideForClass(psiClass)
                val methodSide = SideOnlyUtil.checkElementInMethod(variable)

                var classAnnotated = false

                if (containingClassSide !== Side.NONE && containingClassSide !== Side.INVALID) {
                    if (variableSide !== containingClassSide) {
                        registerVariableError(
                            variable,
                            Error.VAR_CROSS_ANNOTATED_CLASS,
                            variableSide.annotation,
                            containingClassSide.annotation,
                            variableClass
                        )
                    }
                    classAnnotated = true
                }

                if (methodSide === Side.INVALID) {
                    return
                }

                if (variableSide !== methodSide) {
                    if (methodSide === Side.NONE) {
                        if (!classAnnotated) {
                            registerVariableError(
                                variable,
                                Error.VAR_UNANNOTATED_METHOD,
                                variableSide.annotation,
                                methodSide.annotation,
                                variableClass
                            )
                        }
                    } else {
                        registerVariableError(
                            variable,
                            Error.VAR_CROSS_ANNOTATED_METHOD,
                            variableSide.annotation,
                            methodSide.annotation,
                            variableClass
                        )
                    }
                }
            }
        }
    }

    enum class Error {
        VAR_CROSS_ANNOTATED_CLASS {
            override fun getErrorString(vararg infos: Any): String {
                return "A local variable whose class is annotated with ${infos[0]} " +
                    "cannot be used in a class annotated with ${infos[1]}"
            }
        },
        VAR_CROSS_ANNOTATED_METHOD {
            override fun getErrorString(vararg infos: Any): String {
                return "A local variable whose class is annotated with ${infos[0]} " +
                    "cannot be used in a method annotated with ${infos[1]}"
            }
        },
        VAR_UNANNOTATED_METHOD {
            override fun getErrorString(vararg infos: Any): String {
                return "A local variable whose class is annotated with ${infos[0]} " +
                    "cannot be used in an un-annotated method."
            }
        };

        abstract fun getErrorString(vararg infos: Any): String
    }
}
